SwissTransferFeatures/UploadProgressView/TransferSessionManager.swift

```diff
@@ -104,58 +92,116 @@ class TransferSessionManager: ObservableObject {
     }
+}
+
+struct TransferManagerWorker {
+    private static let maxParallelUploads = 4
+    private let uploadURLSession = URLSession.shared
+
+    private let rangeProviderConfig = RangeProvider.Config(
+        chunkMinSize: 50 * 1024 * 1024,
+        chunkMaxSizeClient: 50 * 1024 * 1024,
+        chunkMaxSizeServer: 50 * 1024 * 1024,
+        optimalChunkCount: 200,
+        maxTotalChunks: 10000,
+        minTotalChunks: 1
+    )
+
+    let overallProgress: Progress
 
-    private func uploadFile(atPath: String, toRemoteFile: any RemoteUploadFile, uploadUUID: String) async throws {
-        guard let fileURL = URL(string: atPath) else {
-            throw ErrorDomain.invalidURL(rawURL: atPath)
+    func uploadFile(atPath: String, remoteUploadFileUUID: String, uploadUUID: String) async throws {
+        guard let fileURL = URL(string: atPath),
+              let chunkReader = ChunkReader(fileURL: fileURL) else {
+            throw TransferSessionManager.ErrorDomain.invalidURL(rawURL: atPath)
         }
 
-        let rangeProvider = RangeProvider(fileURL: fileURL, config: TransferSessionManager.rangeProviderConfig)
+        let rangeProvider = RangeProvider(fileURL: fileURL, config: rangeProviderConfig)
 
         let ranges = try rangeProvider.allRanges
-        guard let chunkProvider = ChunkProvider(fileURL: fileURL, ranges: ranges) else {
-            throw ErrorDomain.invalidRangeCompute
-        }
 
-        let rangeCount = ranges.reduce(0) { $0 + $1.count }
-        let fileProgress = Progress(totalUnitCount: Int64(rangeCount))
-        overallProgress?.addChild(fileProgress, withPendingUnitCount: Int64(rangeCount))
-
-        var index: Int32 = 0
-        while let chunk = chunkProvider.next() {
-            guard let rawChunkURL = try injection.sharedApiUrlCreator.uploadChunkUrl(
-                uploadUUID: uploadUUID,
-                fileUUID: toRemoteFile.uuid,
-                chunkIndex: index,
-                isLastChunk: index == ranges.count - 1
-            ) else {
-                throw ErrorDomain.invalidUploadChunkURL
+        try await ranges
+            .dropLast()
+            .enumerated()
+            .map { ($0, $1) }
+            .concurrentForEach(customConcurrency: 4) { index, range in
+                guard let chunk = try chunkReader.readChunk(range: range) else {
+                    throw TransferSessionManager.ErrorDomain.invalidChunk
+                }
+
+                try await uploadChunk(
+                    chunk: chunk,
+                    index: index,
+                    isLastChunk: false,
+                    remoteUploadFileUUID: remoteUploadFileUUID,
+                    uploadUUID: uploadUUID
+                )
             }
 
-            guard let chunkURL = URL(string: rawChunkURL) else {
-                throw ErrorDomain.invalidURL(rawURL: rawChunkURL)
-            }
+        guard let chunk = try chunkReader.readChunk(range: ranges[ranges.count - 1]) else {
```

Comment:
If earlier, you use a `guard let lastChunkRange = ranges.popLast() …` you can do this instead here.

```suggestion
        guard let chunk = try chunkReader.readChunk(range: lastChunkRange) else {
```